### ISSUE TITLE
second darkroom window scaling depending on main screen resolution 

### ISSUE THREAD

**Issue by MStraeten** [darktable-org/darktable#7910]
if the main darktable window is opened on a retina notebook display and the second window is opened on a external display with lower resolution, the second window just displays the upper left quarter of the image (red frame).

![image](https://user-images.githubusercontent.com/39386816/104852465-75200380-58fb-11eb-9362-5a7174eedf32.png)

If the main darktable is opened on the external display (or moved there) and the second window is opened on the higher resolution retina window - the whole image is displayed (the preview window must be opened after the main window is moved) 

**Platform (please complete the following information):**
 - Darktable Version: darktable 3.5.0+705~g76ea0bad0
 - OS: OSX
 - integrated display: 3072x1920 
 - external display: 2560x1440

**Comment by parafin** [comment]
Do you mean that second window is larger than the screen, or that its contents are cropped? I don't know how that second window is rendered, but it's strange that PPD has this effect on it.

**Comment by MStraeten** [comment]
my integrated screen is 16" the external 27" 
The behaviour is independent of the size of the second window - even if it's full screen only the upper left quarter is displayed.
here a sample how it's rendered   
![image](https://user-images.githubusercontent.com/39386816/104853227-b61a1700-58ff-11eb-9d6f-f51a0424b901.png)

**Comment by parafin** [comment]
OK, got it. Probably some simple bug. Not promising to take a look yet though, if anyone else is up to it - go ahead.

**Comment by jenshannoschwalm** [comment]
Yeah, that's clearly related to the ppd / ppd_thb handling for the second window.

**Comment by jenshannoschwalm** [comment]
Found time to have a look at it. In

https://github.com/darktable-org/darktable/blob/688c91aa24a9061268a36da3719c0fe66bfa6634/src/views/darkroom.c#L4492-L4500

we get the ppd via

https://github.com/darktable-org/darktable/blob/688c91aa24a9061268a36da3719c0fe66bfa6634/src/gui/gtk.c#L1568-L1578

and this gets down to

https://github.com/darktable-org/darktable/blob/688c91aa24a9061268a36da3719c0fe66bfa6634/src/osx/osx.mm#L50-L64

@parafin i think at least `NSScreen *nsscreen = [NSScreen mainScreen]` looks not correct ... No chance to check & fix here

**Comment by parafin** [comment]
Wrong screen ppd shouldn't produce such results. We even have ppd override setting in config file - changing it shouldn't produce scaling problems in GUI. Basically wrong ppd setting should have only one of the 2 effects:

1. reduced performance if ppd is too high, due to more pixels being processed than needed;
2. pixelated or blurry image if ppd is too low.

Behaviour described in this ticket is neither of those, so it's a different bug somewhere in how the image is rendered. I suggest trying to force this situation in code by patching _second_window_configure_ppd_dpi to assume ppd==1 and forcing main ppd to 2 through config file. Also consider what happens if window is moved from one display to another.

What you found is indeed a bug on macOS, but a minor one - if second screen has higher ppd, user can just set this ppd in config file, the other way around will only reduce performance. So it's not something breaking things significantly.

**Comment by jenshannoschwalm** [comment]
I agree on most of your arguments. 

> Behaviour described in this ticket is neither of those, so it's a different bug somewhere in how the image is rendered

it think there is a mismatch between found ppd and the size of the rendered image. In case the monitor ppd is 1 but somehow while preparing the image buffer we assumed it to be 2 we would end up at exactly what is reported here.

>  Also consider what happens if window is moved from one display to another

If both displays have different ppds, tricky ...

**Comment by parafin** [comment]
Looking at second_window_expose function in darkroom.c I see at least one problem - it uses dt_cairo_* functions, which take ppd for main window, not second window. But in case of macOS, where this bug is reported, it shouldn't actually matter, since ppd is reported the same. @MStraeten, do you have by any chance ppd set to some other value manually? Then it might not match, since code getting ppd for second window doesn't look and preferences (which might be considered also a bug).

The point is that I don't think that second window code was thought through or/and tested in regard of correct usage of ppd. Introduction of ppd_thb (ui/performance setting) complicated thing further. Someone should check all possible situation in regard to ppd settings. One doesn't have to have high-dpi screen for it, you can just force some value by patching the code. As I said, any combinations shouldn't result in the behaviour described in this bug. I also doubt it's macOS-specific behaviour.

**Comment by MStraeten** [comment]
my darktablerc doesn't contain any ppd values (even after renaming it and got a fresh one created by darktable)
there'S just this line:`screen_dpi_overwrite=-1,000000`

there are these constellations:
integrated display: 3072x1920
external display: 2560x1440
main on integrated display, 2nd window on external: --> upper left quarte is displayed
main on integrated display, 2nd window also on integrated: --> upper left quarte is displayed
main on external display, 2nd window on integrated: --> full image is displayed
main on external display, 2nd window also on external: --> full image is displayed

**Comment by MStraeten** [comment]
adding `screen_ppd_overwrite=96` or modifying `screen_dpi_overwrite=-1` to `screen_dpi_overwrite==95` doesn't change the behaviour

**Comment by parafin** [comment]
dpi is probably not involved. Please try ppd overwrite 1 and 2

**Comment by jenshannoschwalm** [comment]
> I see at least one problem - it uses dt_cairo_* functions, which take ppd for main window, not second window

Right, that's definitely a bug.

> The point is that I don't think that second window code was thought through or/and tested in regard of correct usage of ppd

If that is a hint on me, yes you are right :-)

**Comment by MStraeten** [comment]
same behaviour as reported also with screen_ppd_overwrite=1 or 2

**Comment by parafin** [comment]
> > The point is that I don't think that second window code was thought through or/and tested in regard of correct usage of ppd
> 
> If that is a hint on me, yes you are right :-)

I don't remember/didn't look who added a second window feature:) This stuff is confusing. Or rather was made confusing by Apple and then cairo/gtk + qt joined in the fun.

**Comment by jenshannoschwalm** [comment]
ok, at least we now have at least one definite bug (in the expose section) that **must** be corrected so let's see ...

**Comment by parafin** [comment]
I tried several settings combinations with just one screen and couldn't reproduce this bug on macOS. So it seems to be specific to 2-screen setup, and might also be macOS-specific, or even specific to particular macOS version... Might be a bug somewhere in cairo/gtk...
@MStraeten, am I right to understand that integrated display is HighDPI, while external isn't? What if you change resolution on integrated display to some non-HighDPI mode (don't remember if it's possible)?

**Comment by jenshannoschwalm** [comment]
@MStraeten could you check this in `darkroom.c`  `second_window_expose` ~line 4170

```
  if(image_surface_width != width || image_surface_height != height || image_surface == NULL)
  {
    // create double-buffered image to draw on, to make modules draw more fluently.
    image_surface_width = width;
    image_surface_height = height;
    if(image_surface) cairo_surface_destroy(image_surface);
    image_surface = cairo_image_surface_create(CAIRO_FORMAT_RGB24, width * dev->second_window.ppd, height * dev->second_window.ppd);
    cairo_surface_set_device_scale(image_surface, dev->second_window.ppd, dev->second_window.ppd);
    image_surface_imgid = -1; // invalidate old stuff
  }
  cairo_surface_t *surface;
  cairo_t *cr = cairo_create(image_surface);

  if(dev->preview2_pipe->output_backbuf && // do we have an image?
    dev->preview2_pipe->backbuf_scale == backbuf_scale && // is this the zoom scale we want to display?
    dev->preview2_pipe->backbuf_zoom_x == zoom_x && dev->preview2_pipe->backbuf_zoom_y == zoom_y)
  {
    // draw image
    mutex = &dev->preview2_pipe->backbuf_mutex;
    dt_pthread_mutex_lock(mutex);
    float wd = dev->preview2_pipe->output_backbuf_width;
    float ht = dev->preview2_pipe->output_backbuf_height;
    const int stride = cairo_format_stride_for_width(CAIRO_FORMAT_RGB24, wd);
    surface = cairo_image_surface_create_for_data(dev->preview2_pipe->output_backbuf, CAIRO_FORMAT_RGB24, wd, ht, stride);
    // cairo_surface_set_device_scale(surface, dev->second_window.ppd, dev->second_window.ppd);
```

I think the last line is not good but could also be tried ...

Not looked into "special effects" like focus_peaking ...

**Comment by jenshannoschwalm** [comment]
I can confirm that this is not valid in a one-screen setting testing with different ppd either presented via os setup (via gnome) or dt overriding ppd on linux ...

**Comment by MStraeten** [comment]
removing the comment in the last line is necessary to get a full preview - very laggy and weird focus peaking ...

**Comment by MStraeten** [comment]
no longer reproducible

**Comment by MStraeten** [comment]
reproducible if darktable is restartet after changing main and preview window.

**Comment by MStraeten** [comment]
was fixed via #8870
